Common/Python/PythonInitializer.cs

```diff
@@ -59,30 +61,83 @@ public static void Initialize()
+            // Filter out any paths that are already on our Python path
+            if (paths.IsNullOrEmpty())
             {
                 return;
             }
 
+            // Add these paths to our pending additions
+            _pendingPathAdditions.AddRange(paths);
+
             if (_isInitialized)
             {
                 using (Py.GIL())
                 {
-                    _pendingPathAdditions.AddRange(paths);
+                    dynamic sys = Py.Import("sys");
+                    var locals = new PyDict();
+                    locals.SetItem("sys", sys);
+
+                    // Filter out any already paths that already exist on our current PythonPath
+                    var currentPath = PythonEngine.Eval("sys.path", null, locals.Handle).As<List<string>>();
+                    _pendingPathAdditions = _pendingPathAdditions.Where(x => !currentPath.Contains(x.Replace('\\', '/'))).ToList();
+
+                    // Insert any pending path additions
+                    if (!_pendingPathAdditions.IsNullOrEmpty())
+                    {
+                        var code = string.Join(";", _pendingPathAdditions
+                            .Select(s => $"sys.path.insert(0, '{s}')")).Replace('\\', '/');
+                        PythonEngine.Exec(code, null, locals.Handle);
 
-                    // Generate the python code to add these to our path and execute
-                    var code = string.Join(";", _pendingPathAdditions.Select(s => $"sys.path.append('{s}')"))
-                        .Replace('\\', '/');
+                        _pendingPathAdditions.Clear();
+                    }
 
-                    PythonEngine.Exec($"import sys;{code}");
-                    _pendingPathAdditions.Clear();
+                    locals.Dispose();
                 }
             }
+        }
+
+        /// <summary>
+        /// "Activate" a virtual Python environment by prepending its library storage to Pythons
+        /// path. This allows the libraries in this venv to be selected prior to our base install.
+        /// Requires PYTHONNET_PYDLL to be set to base install.
+        /// </summary>
+        /// <remarks>If a module is already loaded, Python will use its cached version first
+        /// these modules must be reloaded by reload() from importlib library</remarks>
+        public static void ActivatePythonVirtualEnvironment(string pathToVirtualEnv)
+        {
+            if (pathToVirtualEnv == null)
+            {
+                return;
+            }
+
+            if(!Directory.Exists(pathToVirtualEnv)){
+                Log.Error($"PythonIntializer.ActivatePythonVirtualEnvironment(): Path {pathToVirtualEnv} to virtual environment does not exist");
+                return;
+            }
+
+            pathToVirtualEnv = pathToVirtualEnv.TrimEnd('/').TrimEnd('\\');
+            var pathsToPrepend = new List<string>();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
```

Comment:
> OSPlatform.Linux

Ah wonder what about mac/OSX? 

Suggest adding a log in both `if/else` stating the actual final path added?

Reply:
Mac appears to be the same structure as Linux 👍🏽  Adding it to the if case, also added the log 👍🏽